File: backend/app/models/inventory.py
import uuid
from datetime import datetime

from sqlalchemy import CheckConstraint, DateTime, Enum, ForeignKey, Integer, String, Text, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.enums import DispatchStatus, LocationType, RequestStatus, TransactionType
from app.db.base import Base
from app.models.mixins import TimestampMixin, UUIDPrimaryKeyMixin


class InventoryBalance(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "inventory_balances"
    __table_args__ = (
        CheckConstraint("quantity >= 0"),
        CheckConstraint("reserved_quantity >= 0"),
        CheckConstraint("reserved_quantity <= quantity"),
        UniqueConstraint("product_id", "location_type", "location_id", name="uq_inventory_balances_product_location"),
    )

    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"), index=True)
    location_type: Mapped[LocationType] = mapped_column(Enum(LocationType, name="location_type"), index=True)
    location_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), index=True)
    quantity: Mapped[int] = mapped_column(Integer, default=0)
    reserved_quantity: Mapped[int] = mapped_column(Integer, default=0)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class InventoryTransaction(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "inventory_transactions"
    __table_args__ = (CheckConstraint("quantity > 0"),)

    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"), index=True)
    transaction_type: Mapped[TransactionType] = mapped_column(Enum(TransactionType, name="transaction_type"))
    quantity: Mapped[int] = mapped_column(Integer)
    from_location_type: Mapped[LocationType | None] = mapped_column(Enum(LocationType, name="location_type"))
    from_location_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    to_location_type: Mapped[LocationType | None] = mapped_column(Enum(LocationType, name="location_type"))
    to_location_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    reference_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), index=True)
    reference_type: Mapped[str | None] = mapped_column(String(80), index=True)
    notes: Mapped[str | None] = mapped_column(Text)
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), index=True)


class StockMovement(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "stock_movements"
    __table_args__ = (CheckConstraint("quantity > 0"),)

    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"))
    quantity: Mapped[int] = mapped_column(Integer)
    from_location_type: Mapped[LocationType | None] = mapped_column(Enum(LocationType, name="location_type"))
    from_location_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    to_location_type: Mapped[LocationType] = mapped_column(Enum(LocationType, name="location_type"))
    to_location_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True))
    status: Mapped[str] = mapped_column(String(40), default="PENDING")
    reference_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    reference_type: Mapped[str | None] = mapped_column(String(80))
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class AllocationRequest(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "allocation_requests"
    __table_args__ = (CheckConstraint("quantity > 0"),)

    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"), index=True)
    requested_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    approved_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    warehouse_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("warehouses.id"), index=True)
    hub_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("hubs.id"), index=True)
    quantity: Mapped[int] = mapped_column(Integer)
    approved_quantity: Mapped[int | None] = mapped_column(Integer)
    reviewed_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    review_notes: Mapped[str | None] = mapped_column(Text)
    reviewed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    status: Mapped[RequestStatus] = mapped_column(
        Enum(RequestStatus, name="request_status"),
        default=RequestStatus.PENDING,
        index=True,
    )
    notes: Mapped[str | None] = mapped_column(Text)


class DispatchOrder(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "dispatch_orders"
    __table_args__ = (CheckConstraint("quantity > 0"),)

    allocation_request_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("allocation_requests.id"))
    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"))
    dispatched_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    quantity: Mapped[int] = mapped_column(Integer)
    status: Mapped[DispatchStatus] = mapped_column(
        Enum(DispatchStatus, name="dispatch_status"),
        default=DispatchStatus.DRAFT,
    )
    from_location_type: Mapped[LocationType] = mapped_column(Enum(LocationType, name="location_type"))
    from_location_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True))
    to_location_type: Mapped[LocationType] = mapped_column(Enum(LocationType, name="location_type"))
    to_location_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True))
    dispatched_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))


class Receipt(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "receipts"
    __table_args__ = (CheckConstraint("quantity_received > 0"),)

    dispatch_order_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("dispatch_orders.id"))
    product_batch_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("product_batches.id"))
    received_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    quantity_received: Mapped[int] = mapped_column(Integer)
    notes: Mapped[str | None] = mapped_column(Text)
    received_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class AgentAllocation(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "agent_allocations"
    
    agent_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("agents.id"))
    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"))
    quantity: Mapped[int] = mapped_column(Integer)
    status: Mapped[str] = mapped_column(String, default="PENDING")
    allocated_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    handed_over_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))

class AgentSale(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "agent_sales"
    
    agent_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("agents.id"))
    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id"))
    quantity: Mapped[int] = mapped_column(Integer)
    recorded_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))

File: frontend/src/types/inventory.ts
export type Product = {
  id: string;
  name: string;
  sku: string;
  description: string | null;
  unit: string;
  low_stock_threshold: number;
  is_active: boolean;
  created_at: string;
};

export type ProductPage = {
  items: Product[];
  total: number;
  limit: number;
  offset: number;
};

export type ProductBatch = {
  id: string;
  product_id: string;
  manufacturer_id: string;
  batch_number: string;
  quantity: number;
  status: "DRAFT" | "RELEASED_TO_WAREHOUSE" | "RECEIVED_AT_WAREHOUSE" | "CANCELLED";
  produced_at: string;
  released_at: string | null;
  received_at: string | null;
  created_at: string;
};

export type WarehouseRecord = {
  id: string;
  name: string;
  location: string | null;
  manager_id: string | null;
  is_active: boolean;
  created_at: string;
};

export type HubRecord = {
  id: string;
  name: string;
  location: string | null;
  warehouse_id: string;
  manager_id: string | null;
  is_active: boolean;
  created_at: string;
};

export type AllocationRequest = {
  id: string;
  product_id: string;
  requested_by: string;
  approved_by: string | null;
  warehouse_id: string | null;
  hub_id: string | null;
  quantity: number;
  approved_quantity: number | null;
  reviewed_by: string | null;
  review_notes: string | null;
  reviewed_at: string | null;
  status: "PENDING" | "APPROVED" | "REJECTED" | "FULFILLED" | "CANCELLED";
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type DispatchOrder = {
  id: string;
  allocation_request_id: string | null;
  product_id: string;
  dispatched_by: string | null;
  quantity: number;
  status: "DRAFT" | "DISPATCHED" | "PARTIALLY_RECEIVED" | "RECEIVED" | "CANCELLED";
  from_location_type: string;
  from_location_id: string;
  to_location_type: string;
  to_location_id: string;
  dispatched_at: string | null;
  created_at: string;
  updated_at: string;
};

export type InventoryBalance = {
  id: string;
  product_id: string;
  location_type: "MANUFACTURER" | "WAREHOUSE" | "HUB" | "AGENT";
  location_id: string;
  quantity: number;
  reserved_quantity: number;
  updated_at: string;
};

export type InventoryTransaction = {
  id: string;
  product_id: string;
  transaction_type: string;
  quantity: number;
  from_location_type: string | null;
  from_location_id: string | null;
  to_location_type: string | null;
  to_location_id: string | null;
  reference_id: string | null;
  reference_type: string | null;
  notes: string | null;
  created_by: string;
  created_at: string;
};

export type AgentRecord = {
  id: string;
  name: string;
  phone: string | null;
  hub_id: string;
  agent_code: string;
  is_active: boolean;
};

export type AgentAllocationRecord = {
  id: string;
  agent_id: string;
  product_id: string;
  quantity: number;
  status: "PENDING" | "HANDED_OVER";
  created_at: string;
};